[Sema] Only require a default implementation for SPI requirements in non-SPI protocols

xymus opened the pull request:
The check for SPI protocol requirements needing a default implementation should ignore requirements in a protocol that is SPI too. The protocol can’t be conformed to without knowing of the requirement.

rdar://65286171

xymus committed:
[Sema] Move the check for SPI protocol requirements to existing logic

Asides from removing duplicated code this change also now checks implicit SPI
requirements and applies only in library evolution mode.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

beccadax reviewed (approved):
This change looks good. However, is there a runtime test verifying that test cases like `PublicProto` work correctly?

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -876,39 +876,41 @@ void AttributeChecker::visitSPIAccessControlAttr(SPIAccessControlAttr *attr) {
-              extension->isConstrainedExtension())
-            return false;
+      if (!protocol->isSPI()) {
+        auto implementations = TypeChecker::lookupMember(
```
This should reuse the existing logic for finding default implementations instead of rolling its own.

slavapestov replied:
More specifically I think these checks could be moved to typechecker::inferDefaultWitnesses().

xymus replied:
Thank you for pointing this out. I moved the check and it is much cleaner!

xymus committed:
[Sema] SPI requirements in SPI protocols don't need a default implementation

The check for SPI protocol requirements needing a default implementation
should ignore requirements in a protocol that is SPI too. The protocol
can’t be conformed to without knowing of the requirement.

rdar://65286171

xymus committed:
[Sema] Add test case for SPI requirements on name vs type

xymus commented:
Moved the check to TypeCheckProtocol with pre-existing logic.

I'll have to spend more time on the runtime test. It works as expected with library evolution but it crashes in an ugly way without it when rebuilding from the public swiftinterface.

@swift-ci Please smoke test
